### DISCUSSION TITLE
Different inference results for Detectron2DetectionModel than plain detectron

### DISCUSSION THREAD

**Discussion by ninafiona** [obss/sahi#644] [Q&A] [answered]
I ran inference using SAHI (without slicing) and with my model using just detectron and got different results. I did this for validation / debugging. Here is my code:

SAHI version:
```
cfg_path = '..'
checkpoint_path = '...'
image_path = '...'
detection_model = Detectron2DetectionModel(
  model_path=checkpoint_path,
  config_path=cfg_path
)
image = cv2.imread(image_path)
prediction_result = get_prediction(
  image=image,
  detection_model=detection_model
)
```

Plain Detectron version:
```
cfg = get_cfg()
cfg.merge_from_file(cfg_path)
model = build_model(cfg) 
DetectionCheckpointer(model).load(checkpoint_path)
image = Image.open(input_image_path)
convert_tensor = transforms.ToTensor()
img_tensor = convert_tensor(image) * 255
inputs = [dict(image=img_tensor)]
model.eval()
with torch.no_grad():
  outputs = model(inputs)
```

The bounding boxes are different (just 4 boxes prediced by SAHI model and 10 by second). I don't apply any postprocessing (nonmax etc.).

**Comment by fcakyon** [comment, marked as answer]
@ninafiona thanks for your interest in `SAHI`!

SAHI follows [DefaultPredictor implementation](https://github.com/obss/sahi/blob/2bf0cbd804553f97b92562488ab2fed08daed74a/sahi/model.py#L566)  which [applies image resizing by default](https://detectron2.readthedocs.io/en/latest/modules/engine.html#detectron2.engine.defaults.DefaultPredictor).

A difference between `build_model` and `DefaultPredictor` is expected and by following [this discussion](https://github.com/facebookresearch/detectron2/discussions/3097) `DefaultPredictor` implementation is better in our case since it automatically does all the preprocessing.

In your example, if you resize the image to default sizes before giving it to the model as input, the results will be similar.

**Comment by ninafiona** [comment]
Thanks, it turns out two aspects were the cause:
* even though `cfg.INPUT.MAX_SIZE_TEST` was defined, the default predictor resized it to another value for some reason so I made some adjustments to resize it to the same size
* numpy reads images as RGB and PIL as BGR and that also caused a difference

Now I get the same outputs for both approaches!
